sphinx/texinputs/sphinx.sty

```diff
@@ -683,8 +685,29 @@
   \sphinxsetvskipsforfigintablecaption
   \begin{minipage}{#1}%
 }{\end{minipage}}
-% store original \caption macro for use with figures in longtable and tabulary
-\AtBeginDocument{\let\spx@originalcaption\caption}
+% store the original \caption macro for usage with figures inside longtable
+% and tabulary cells. Make sure we get the final \caption in presence of
+% caption package, whether the latter was loaded before or after sphinx.
+\AtBeginDocument{%
+    \let\spx@originalcaption\caption
+    \@ifpackageloaded{caption}
+      {\let\spx@ifcaptionpackage\@firstoftwo
+       \caption@AtBeginDocument*{\let\spx@originalcaption\caption}%
+% in presence of caption package, drop our own \sphinxcaption whose aim was to
+% ensure same width of caption to all kinds of tables (tabular(y), longtable),
+% because caption package has its own width (or margin) option
+       \def\sphinxcaption{\caption}%
+% vertical spacing from last baseline of a top caption to top rule of
+% table will still be \sphinxbelowcaptionspace for tabular/tabulary,
+% (i.e. \sphinxaftertopcaption needs no modification) but we now need
+% this definition for the corrective skip inserted by longtable template:
+       \def\sphinxlongtablecapskipadjust
+          {\dimexpr-\dp\strutbox-\abovecaptionskip+
+                   \sphinxbelowcaptionspace\relax}% this is the vspace we want
+      }%
+      {\let\spx@ifcaptionpackage\@secondoftwo}%
+}
```

Comment:
The LaTeX code comments explain what this does. This is compatibility layer with caption package: if detected, do not use `\sphinxcaption` own code at all and slightly adjust corrective measure for longtable because `caption` package patches longtable code and makes it use an other vertical spacing.

Some other modifications of behaviour in presence of `caption` package are implemented in the `sphinxVerbatim` environement and use the `\spx@ifcaptionpackage` conditional defined here.